### PR TITLE
fix tightbbox to account for markeredgewidth

### PR DESCRIPTION
**Pull request by brunobeltran**
## PR Summary

This initial PR is a naive fix for issue #16606; it will be updated as I document what the expected behavior should be for each marker type. 

In short, increases the padding added to `Line2D.get_window_extent` in order to account for markers with non-trivial edge widths.

I also do not know what kind of test would be appropriate to add here? Just code that makes a plot that's composed of one marker with known size and edge width and then check that `get_tightbbox` returns the right values? Is the goal one test per marker type? 

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/lines.py [hunk 1]
```diff
@@ -617,7 +617,8 @@ def get_window_extent(self, renderer):
                                  ignore=True)
         # correct for marker size, if any
         if self._marker:
-            ms = (self._markersize / 72.0 * self.figure.dpi) * 0.5
+            extra_pts = self._markersize + self._markeredgewidth
```
Comment (timhoffm) [inline lib/matplotlib/lines.py]:
```suggestion
            extra_pts = self._markersize + self._markeredgewidth / 2
```

I think. The edge line is drawn centered on the marker outline. So only half of the width adds to the extent.

Comment (brunobeltran) [reply lib/matplotlib/lines.py]: You're right, but this is already accounted for in the `* 0.5` for both marker size and edge width in the next line. In the more general case, this multiplication is no longer needed, see newest commits.

**Comment by brunobeltran** [comment]
Some design decisions that I didn't know how to make:

1) I implemented a `get_centered_bbox` function. Do I implement `get_extents`, or `get_bbox`, or something else instead? There doesn't seem to be an agreed-upon convention in the codebase for how to extract the bbox information from an object (lines have `get_window_extents`, paths have `get_extents`, etc.)
2) the information to calculate the `PathEndAngles` for each glyph can *theoretically* be extracted from the path itself. I opted to manually tabulate them instead since the number of glyphs is small and they are geometrically straightforward. A more generalized approach that identifies the points on the path that "create" the bbox, then extracts their PathEndAngles from the path itself would be more general and theoretically resilient to future/custom glyph types, but would be quite a bit more work.
3) more in general, I'm not familiar enough with the mpl codebase as a whole to know if there's already a convention for aggragating data like I do using BoxSides(PathEndAngles(...))
4) On that note, as long as we don't do something like suggested above in (2), we will now be storing more than two per-glyph-type data things (and if e.g. #15703 gets implemented we'll have a couple of more pieces of per-glyph-type information, the area and visual scaling constants, saved as well). This might argue for a simple refactor of the current `markers` dictionary into a list of `GlyphInfo` objects, containing (for each glyph) the 
    a) list of valid short names (e.g. ['', ' ', 'None', None])
    b) long name (e.g. 'star')
    c) PathEndAngles (or whatever they end up being called)
    d) area of glyph, to scale with
    e) (in the future) effective visual size scaling constant

I'm happy to do that simple refactor here, since I'm adding the information, but wanted to make sure there was a shared feeling that it is necessary in the first place, and that I wouldn't be wasting too much time doing it in a style that would immediately be rejected.

**Comment by brunobeltran** [comment]
> > Just code that makes a plot that's composed of one marker with known size and edge width and then check that get_tightbbox returns the right values?
> 
> I don‘t know if it‘s necessary/worth to distinguish different markers.

Now that it's more obvious that this gets messy on a per-glyph basis, should we include one test per glyph or just test a couple of the nasty ones? (say, 'p', '*', 'o', 'x', to get examples of glyphs with miter joins, round joins, filled and not)?

**Comment by brunobeltran** [comment]
Weird. The tests I just added pass in a Jupyter notebook but not with the Agg backend or on Travis?

Basically, the code
```
_draw_marker_outlined('*', markeredgewidth=20)
```
produces a 480x480px image with the box in the wrong place. 

marker_bbox_star

The calling the identical function in Jupyter...
```
test_marker._draw_marker_outlined('*', markeredgewidth=20)
plt.savefig('marker_bbox_star.png')
```
produces a correct figure, but 345x345px. 
marker_bbox_star


I don't think this is a problem in my own code, since the box is incorrect even for the trivial case of `markeredgewidth=0`, and doing the following in Jupyter produces a 480x480px image that is also correct
```
test_marker._draw_marker_outlined('*', markeredgewidth=20)
plt.savefig('marker_bbox_star.png', dpi=100)
```
marker_bbox_star-expected

**Comment by brunobeltran** [comment]
@timhoffm Ready for re-review and need help.

1) My new tests are passing in Jupyter, but not in the console (see above).
2) I ended up adding a new public API element `MarkerStyle.get_centered_bbox`, but don't know where or how to document it appropriately besides its docstring.

**Comment by timhoffm** [comment]
I think this is a dpi issue. Some calculation does not Takt them into account. The jupyter backend (%matplotlib inline) sets 72dpi by default, everything else uses 100dpi. What happens if you do ˋplt.rcParams[„figure.dpi“]ˋ?

Sorry for the formatting, I‘m on mobile.

I‘m quite busy right now. Please be patient with a full review. That may take a couple of days.

**Comment by brunobeltran** [comment]
Thanks for the quick reply Tim! Looks like it is a DPI thing, I'll follow that lead.

Of course I assume you're very busy, and I definitely wasn't meaning to rush you. I just wanted to ping so you knew it was safe to start looking without wasting even more of your time.

Thanks in advance for your help!

**Comment by ImportanceOfBeingErnest** [comment]
You removed the line `ms = (self._markersize / 72.0 * self.figure.dpi) * 0.5` which accounts for dpi. So the box will be wrong for every dpi setting, except `72`, where `1/72*dpi==1`.

**Comment by brunobeltran** [comment]
Haha, didn't see the "member" when I replied to you earlier @ImportanceOfBeingErnest. Hopefully I didn't come off as presumptuous!

Mostly, thanks for the good eye! I had forgotten that the markers stuff is in points. Fixed the code to make that consistent. Tests should pass now. Had to push up new "baseline_images", since they were generated at different DPI for whatever reason than before, but now look as expected.

**Comment by ImportanceOfBeingErnest** [comment]
To have image comparisson look like the default style one can use 

    @image_comparison(..., style='mpl20')

...just in case that is still relevant. One could also try to use only a single image (image comparison is rather expensive, so if it can be reduced to a minimum, that would sure be nice)

In general I haven't quite understood how wrong adding half the edge width would be and if that really warrants adding this huge code block. In particular, I would have though adding half the edgewidth would rather *over*estimate the bbox - which would be totally acceptable in my eyes.
Maybe an example case either here, or in the  original issue #16606 would be helpful in that respect.

**Comment by jklymak** [comment]
I agree with @ImportanceOfBeingErnest.  This is for the rare case that a marker over-spills its axes, and most folks don't have markeredgewsidths that are much larger than a point or two.  This is really cool that you sorted through all the geometry, but is it overkill compared to the actual problem?

File: lib/matplotlib/markers.py [hunk 2]
```diff
@@ -142,6 +143,233 @@
...
+        # "butt" for all markers. if we find otherwise, we should change this
+        # code to check for the "projecting" and "round" cases
+        if capstyle != 'butt':
+            raise NotImplementedError("Only capstyle='butt' currently needed")
+        pad = (width/2) * np.cos(phi)
+    # the two "same as straight line" cases are NaN limits in the miter formula
+    elif np.isclose(theta, 0) and np.isclose(phi, 0) \
+    or np.isclose(theta, np.pi) and np.isclose(phi, np.pi/2):
+        pad = width/2
+    # to calculate the offset for _joinstyle == 'miter', imagine aligning the
```
Comment (anntzer) [inline lib/matplotlib/markers.py]: If you're really going to go through all this (I agree with the other maintainers that this looks slightly overkill...), you may also want to take miter_limits into account -- see https://github.com/matplotlib/matplotlib/issues/9830.

**Comment by brunobeltran** [comment]
Thanks everyone for reading through this mess. While a decision that the existing code is "overkill" would be disappointing, I can definitely see the argument and would be happy to check in a simpler approximation. 

See my update to #16606 to see the kinds of use cases where having a more tight approximation is nice (tl;dr: prevents having to manually fix things in illustrator, makes sure things line up nicely). But I agree that an overestimate is "sufficient".

If it was less code, or more generally applicable (for example, adding code to perform real bbox computation for an arbitrary stroked path), would it then be worth it? We can maybe check in the approximate version for now, and I can send a PR for a less complex "exact" solution once I have the generalized stroked path bbox code.

**Comment by anntzer** [comment]
For me I think the main problem is the big list of hardcoded marker properties.  Can they not be inferred from the path of the marker?

Also, if the much shorter solution of slightly overestimating the widths by just shifting by a full markeredgewidth suffices, then I would certainly prefer that.  (I guess that in fact if you set the miter limit correctly (which right now is not exposed, it's really somewhere deep in the rendering stack), you can in fact guarantee that mew (times a factor of your choice) is always enough?)

**Comment by brunobeltran** [comment]
> For me I think the main problem is the big list of hardcoded marker properties. Can they not be inferred from the path of the marker?

If that's the problem, I'm happy to write a slighly more general solution that computes these properties from the path of the marker. It is possible, just a little difficult in the general case as computing the extents of stroked bezier paths is [famously complicated](https://pdfs.semanticscholar.org/9ac1/978746ec54bdd555b906e2ea1eb922cd6ffd.pdf). However, I do know how to do it correctly, and since I use a lot of custom markers, it would greatly benefit me since my fix would then apply in that case as well.

> Also, if the much shorter solution of slightly overestimating the widths by just shifting by a full markeredgewidth suffices, then I would certainly prefer that. (I guess that in fact if you set the miter limit correctly (which right now is not exposed, it's really somewhere deep in the rendering stack), you can in fact guarantee that mew (times a factor of your choice) is always enough?)

If I do this, will it get mainlined? If so I'll do it (in a separate PR presumably?, since it'd be a *very* simple change), and can work on the more general fix you suggest above here.

**Comment by brunobeltran** [comment]
> For me I think the main problem is the big list of hardcoded marker properties. Can they not be inferred from the path of the marker?


Conveniently, the currently hardcoded figure properties mean we would have amazing test coverage of this more general solution with no extra work I suppose....

**Comment by timhoffm** [comment]
Wow, @brunobeltran you really dug into the details. Respect for that! 😄 

From a quick look, i have two concerns:

1. Maintainability. Do we want to guarantee perfect tightbboxes in the future? We have to maintain this detailed code.
2. Performance. Does the exact caclulation slow us down significantly?

That said, I have not yet read the code in detail and thus cannot judge it.

In case we decide a simplified approximation is enough for matplotlib: Maybe there is (or can be introduced) a simple function that can be monkey-patched to inject the exact calculation. This would at least allow @brunobeltran to apply this code for his usecase, even if it's not in core matplotlib.

**Comment by jklymak** [comment]
I sympathize with wanting to get this right, but huge marker edge widths don't seem very common or desirable so I am still would prefer a simple "close-enough" solution.

That said,your solution is relatively self contained, so maybe its OK to go in.  The only possible objection would be if its a huge performance drop.  Have we timed this on a "normal" scatter plot with many entries?

**Comment by anntzer** [comment]
> If that's the problem, I'm happy to write a slighly more general solution that computes these properties from the path of the marker. It is possible, just a little difficult in the general case as computing the extents of stroked bezier paths is famously complicated. However, I do know how to do it correctly, and since I use a lot of custom markers, it would greatly benefit me since my fix would then apply in that case as well.

But I don't think we need to cover the general case here, just polygons and circles.  I guess polygons would be relatively simpler; for circles, somehow luckily, falling back to 0.5*mew would work :-)

> If I do this, will it get mainlined?

Well that's the problem as always: any code that goes in needs to be maintained and is extremely annoying to get rid of, and maintainer time is finite.  That's why we're not necessarily so keen on adding a few hundreds of line of code just to handle a relatively edge case :-)

> Conveniently, the currently hardcoded figure properties mean we would have amazing test coverage of this more general solution with no extra work I suppose....

Not a huge fan of playing the coverage numbers game :-)

**Comment by brunobeltran** [comment]
@anntzer , I got a little swamped with paper deadlines, but started working on code for general solution today (written, just needs testing).

However, I ran into a bit of a design issue. Right now bezier.py imports path.py (but only in a couple of functions that feel to me like they should be in path.py anyway...). Obviously a general routine that calculates bbox for a stroked path belongs as a method of Path. But I need helper routines from bezier.py so....

Should I
1) put my code in bezier.py/markers.py (as in current commit, less elegant, but no API breaking)]
2) fix the underlying issue, move appropriate code from bezier.py into path.py, do it "right" (internal API break, but that code is really only used in one place internally...)

in order to minimize resistance to this already pretty contentious PR?

**Comment by anntzer** [comment]
I would be fine with moving stuff from bezier.py to path.py if that's necessary.
You may consider breaking stuff into multiple consecutive PRs to prevent review from going out of control... ;)

**Comment by timhoffm** [comment]
Before there‘s going more work into this: Are we positive, that we want to take on the exact solution with the added amount of code vs. an approximate solution? It would be a shame to detail it all out and later decide that the maintenance burden would be too high.

I haven‘t looked into this in detail, but the amount of code needed to special-case the exact solution scares me a little.

Ping @jklymak @anntzer @tacaswell

**Comment by anntzer** [comment]
I think this will depend on the amount of code involved...  (I can't guarantee a quick review, but will try to keep a look on this.)
I just realized there may be an easier solution: I *think* the "true" size of a marker likely always grows as `a*markersize+b*markeredgewidth+c`?  in which case we could just get away with estimating `a`, `b` and `c` by rasterizing the marker at a few ms/mew values (the results can be cached per-marker), instead of doing the complicated geometry calculations.

**Comment by brunobeltran** [comment]
Yeah @anntzer's proposal was actually my plan after you guys asked about performance.

I would be adding ~2 small helper functions to bezier.py, ~2 small helper methods to Path, and then the "get_window_extents" function for marker would then just either cache the result on first call or even on module load, since it's a practically instantaneous calculation.

**Comment by brunobeltran** [comment]
Superceded by #17119, a cleaner implementation using some of my recent additions to `Path`.

**Pull request closed without merging**